src/include/optimizer/rule_impls.h

```diff
@@ -0,0 +1,823 @@
+   * @returns Whether the input OperatorExpression passes the check
+   */
+  bool Check(common::ManagedPointer<OperatorExpression> plan, OptimizeContext *context) const override;
+
+  /**
+   * Transforms the input expression using the given rule
+   * @param input Input OperatorExpression to transform
+   * @param transformed Vector of transformed OperatorExpressions
+   * @param context Current OptimizeContext executing under
+   */
+  void Transform(common::ManagedPointer<OperatorExpression> input,
+                 std::vector<std::unique_ptr<OperatorExpression>> *transformed,
+                 OptimizeContext *context) const override;
+};
+
+/**
+ * Rule transforms (A JOIN B) JOIN C -> A JOIN (B JOIN C)
+ */
+
+class InnerJoinAssociativity : public Rule {
+ public:
+  /**
+   * Constructor
+   */
+  InnerJoinAssociativity();
+
+  /**
+   * Checks whether the given rule can be applied
+   * @param plan OperatorExpression to check
+   * @param context Current OptimizeContext executing under
+   * @returns Whether the input OperatorExpression passes the check
+   */
+  bool Check(common::ManagedPointer<OperatorExpression> plan, OptimizeContext *context) const override;
+
+  /**
+   * Transforms the input expression using the given rule
+   * @param input Input OperatorExpression to transform
+   * @param transformed Vector of transformed OperatorExpressions
+   * @param context Current OptimizeContext executing under
+   */
+  void Transform(common::ManagedPointer<OperatorExpression> input,
+                 std::vector<std::unique_ptr<OperatorExpression>> *transformed,
+                 OptimizeContext *context) const override;
+};
+
+//===--------------------------------------------------------------------===//
+// Implementation rules
+//===--------------------------------------------------------------------===//
+
+/**
+ * Rule transforms Logical Scan -> Sequential Scan
+ */
+class GetToSeqScan : public Rule {
+ public:
+  /**
+   * Constructor
+   */
+  GetToSeqScan();
+
+  /**
+   * Checks whether the given rule can be applied
+   * @param plan OperatorExpression to check
+   * @param context Current OptimizeContext executing under
+   * @returns Whether the input OperatorExpression passes the check
+   */
+  bool Check(common::ManagedPointer<OperatorExpression> plan, OptimizeContext *context) const override;
+
+  /**
+   * Transforms the input expression using the given rule
+   * @param input Input OperatorExpression to transform
+   * @param transformed Vector of transformed OperatorExpressions
+   * @param context Current OptimizeContext executing under
+   */
+  void Transform(common::ManagedPointer<OperatorExpression> input,
+                 std::vector<std::unique_ptr<OperatorExpression>> *transformed,
+                 OptimizeContext *context) const override;
+};
+
+/**
+ * Rule transforms LogicalExternalFileGet -> ExternalFileGet
+ */
+class LogicalExternalFileGetToPhysical : public Rule {
```

Comment:
`LogicalExternalFileGetToPhysicalExternalFileGet`

On second thought... that one is really ugly...